From fc8ce2f25a4abea34295d265002c7d368a81be1a Mon Sep 17 00:00:00 2001
From: allen-bst <70563899+allen-bst@users.noreply.github.com>
Date: Fri, 24 Nov 2023 15:53:48 +0800
Subject: [PATCH 090/224] [A13] Case ROB-8853 : allow chrome to install apk
 (#83)

* Case ROB-8853 : allow chrome to install apk
* Case ROB-8853 : adding permissionCode check as to allow only request install packages
---
 .../java/com/android/server/appop/AppOpsService.java | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/services/core/java/com/android/server/appop/AppOpsService.java b/services/core/java/com/android/server/appop/AppOpsService.java
index 20f0c17bb9d9..d2d4444bd0d8 100644
--- a/services/core/java/com/android/server/appop/AppOpsService.java
+++ b/services/core/java/com/android/server/appop/AppOpsService.java
@@ -3233,6 +3233,10 @@ public class AppOpsService extends IAppOpsService.Stub {
         if (resolvedPackageName == null) {
             return AppOpsManager.MODE_IGNORED;
         }
+        /* ROB-8853 allow apps to directly install from chrome*/
+        int permissionCode = permissionToOpCode(Manifest.permission.REQUEST_INSTALL_PACKAGES);
+        if (packageName != null && packageName.equals("com.android.chrome") && (code == permissionCode))
+            return AppOpsManager.MODE_ALLOWED;
         return checkOperationUnchecked(code, uid, resolvedPackageName, attributionTag, raw);
     }
 
@@ -3440,6 +3444,7 @@ public class AppOpsService extends IAppOpsService.Stub {
             return new SyncNotedAppOp(AppOpsManager.MODE_IGNORED, code, attributionTag,
                     packageName);
         }
+
         return noteOperationUnchecked(code, uid, resolvedPackageName, attributionTag,
                 Process.INVALID_UID, null, null, AppOpsManager.OP_FLAG_SELF,
                 shouldCollectAsyncNotedOp, message, shouldCollectMessage);
@@ -3508,7 +3513,12 @@ public class AppOpsService extends IAppOpsService.Stub {
             } else {
                 final Op switchOp = switchCode != code ? getOpLocked(ops, switchCode, uid, true)
                         : op;
-                final int mode = switchOp.evalMode();
+                int mode = switchOp.evalMode();
+                /* ROB-8853 allow apps to directly install from chrome*/
+                int permissionCode = permissionToOpCode(Manifest.permission.REQUEST_INSTALL_PACKAGES);
+                if (packageName != null && packageName.equals("com.android.chrome") && (code == permissionCode))
+                    mode = AppOpsManager.MODE_ALLOWED;
+
                 if (mode != AppOpsManager.MODE_ALLOWED) {
                     if (DEBUG) Slog.d(TAG, "noteOperation: reject #" + mode + " for code "
                             + switchCode + " (" + code + ") uid " + uid + " package "
